src/sknnr_spatial/image/_base.py

```diff
@@ -0,0 +1,193 @@
+        """
+
+    def _get_nodata_mask(self, flat_image: ImageType) -> ImageType | None:
+        """
+        Get a mask of NoData values in the shape (pixels,) for the flat image.
+
+        NoData values are represented by True in the output array.
+        """
+        # Skip allocating a mask if the image is float and nodata wasn't given
+        if not (is_float := flat_image.dtype.kind == "f") and self.nodata_vals is None:
+            return None
+
+        mask = self._backend.zeros(flat_image.shape, dtype=bool)
+
+        # If it's floating point, always mask NaNs
+        if is_float:
+            mask |= self._backend.isnan(flat_image)
+
+        # If nodata was specified, mask those values
+        if self.nodata_vals is not None:
+            mask |= flat_image == self.nodata_vals
+
+        # Set the mask where any band contains nodata
+        return mask.max(axis=self.flat_band_dim)
+
+    def _validate_nodata_vals(
+        self, nodata_vals: float | tuple[float] | NDArray | None
+    ) -> NDArray | None:
+        """
+        Get an array of NoData values in the shape (bands,) based on user input.
+
+        Scalars are broadcast to all bands while sequences are checked against the
+        number of bands and cast to ndarrays. There is no need to specify np.nan as a
+        NoData value because it will be masked automatically for floating point images.
+        """
+        if nodata_vals is None:
+            return None
+
+        # If it's a numeric scalar, broadcast it to all bands
+        if isinstance(nodata_vals, (float, int)) and not isinstance(nodata_vals, bool):
+            return np.full((self.n_bands,), nodata_vals)
+
+        # If it's not a scalar, it must be an interable
+        if not isinstance(nodata_vals, Sized) or isinstance(nodata_vals, (str, dict)):
+            raise TypeError(
+                f"Invalid type `{type(nodata_vals).__name__}` for `nodata_vals`. "
+                "Provide a single number to apply to all bands, a sequence of numbers, "
+                "or None."
+            )
+
+        # If it's an iterable, it must contain one element per band
+        if len(nodata_vals) != self.n_bands:
+            raise ValueError(
+                f"Expected {self.n_bands} nodata values but got {len(nodata_vals)}. "
+                f"The length of `nodata_vals` must match the number of bands."
+            )
+
+        return np.asarray(nodata_vals, dtype=float)
+
+    def _fill_nodata(self, flat_image: ImageType, nodata_fill_value=0.0) -> ImageType:
+        """Fill values in a flat image based on the pre-calculated mask."""
+        if self.nodata_mask is None:
+            return flat_image
+
+        if isinstance(nodata_fill_value, float):
+            flat_image = flat_image.astype(float)
+
+        flat_image[self.nodata_mask, :] = nodata_fill_value
+
+        return flat_image
+
+
+@singledispatch
+def predict(
+    X_image: NDArray | xr.DataArray | xr.Dataset,
```

Comment:
Good point! I thought mypy or ruff would complain about that because `ImageType` is a TypeVar but it doesn't affect the return type for these cases, but apparently not.